[AMD] Added bufferOps refinement

ravil-mobile opened the pull request:
@guacamoleo, I added the buffer ops refinement. I tested it with our GEMM kernel. Numerics is ok but performance dropped by a factor 2. @giuseros, what do you think?

```bash
❯ AMDGCN_USE_BUFFER_OPS=1 TRITON_HIP_STREAM_MAX_DEPTH=1 MLIR_ENABLE_DUMP=0 TRITON_PRINT_AUTOTUNING=0 python3 ./gemm-ex.py -f ./exp-config.yaml --dump-ir ttgir
MASKING load/store: disabled
MATRIX B TRANSPOSED: false
use_bias=False
matmul_kernel
perf: 215.41678697824958 TFLOP/s
✅ Triton and Torch match

❯ AMDGCN_USE_BUFFER_OPS=1 TRITON_HIP_STREAM_MAX_DEPTH=1 MLIR_ENABLE_DUMP=1 TRITON_PRINT_AUTOTUNING=0 python3 ./gemm-ex.py -f ./exp-config.yaml --dump-ir ttgir --trans-b
MASKING load/store: disabled
MATRIX B TRANSPOSED: true
use_bias=False
matmul_kernel
perf: 213.28888766541104 TFLOP/s
✅ Triton and Torch match

❯ AMDGCN_USE_BUFFER_OPS=1 TRITON_HIP_STREAM_MAX_DEPTH=1 MLIR_ENABLE_DUMP=0 TRITON_PRINT_AUTOTUNING=0 python3 ./gemm-ex.py -f ./exp-config.yaml --dump-ir ttgir --trans-a
MASKING load/store: disabled
MATRIX B TRANSPOSED: false
use_bias=False
matmul_kernel
perf: 230.4643251881426 TFLOP/s
✅ Triton and Torch match

❯ AMDGCN_USE_BUFFER_OPS=1 TRITON_HIP_STREAM_MAX_DEPTH=1 MLIR_ENABLE_DUMP=0 TRITON_PRINT_AUTOTUNING=0 python3 ./gemm-ex.py -f ./exp-config.yaml --dump-ir ttgir --trans-a --trans-b
MASKING load/store: disabled
MATRIX B TRANSPOSED: true
use_bias=False
matmul_kernel
perf: 207.3770854286465 TFLOP/s
✅ Triton and Torch match
```

ravil-mobile committed:
[AMD] Added a skeleton for the `RefineOpsPass`

ravil-mobile committed:
[AMD] Added a prototype of convertDot to RefineOps pass

ravil-mobile committed:
[AMD] Added a proto of the ConcatOp to the AMDGPU dialect

ravil-mobile committed:
[AMD] Added ConcatOpToLLVM rewrite pattern

ravil-mobile committed:
[AMD] Added tt.LoadOp and tt.LocalStore refinement

ravil-mobile committed:
[AMD] restructured `refineOps.cpp` and added `reschedule` pass

ravil-mobile committed:
[AMD] moved membar analysis to its dedicated pass

ravil-mobile committed:
[AMD] added `refine-ops` opstion as a sched.hint variant

ravil-mobile committed:
[AMD] Added a proto of the dependency-graph

ravil-mobile committed:
[AMD] refactored dependency-graph builders

ravil-mobile committed:
[AMD] moved ttg.local_load rewrite in refineOps to its own place

ravil-mobile committed:
[AMD] Implemented machine model for the rescheduling pass

guacamoleo committed:
Adding support for dot-tiling

guacamoleo committed:
Adding DotTileAttr to extract_slices of local_loads

guacamoleo committed:
Fixing dot-tiling for different shapes and orders.

guacamoleo committed:
Reformat from clang-format.

ravil-mobile committed:
[AMD] Improved re-scheduling

guacamoleo committed:
Addressing review improvements.

ravil-mobile committed:
Applied clang-format to DotTiling.h and RefineOps.cpp

ravil-mobile committed:
[AMD] Fixed lit-tests

ravil-mobile committed:
[AMD] adjusted calculations coming from calcDotTileShape

ravil-mobile committed:
[AMD] improved canonicalizer for extract and concat ops

ravil-mobile committed:
[AMD] bug-fix in reschedule-ops pass: barriers for local_alloc ops

ravil-mobile committed:
[AMD] Changed machine model op selector in resched. pass

guacamoleo committed:
DotTiling to handle invalid configurations more robustly.

guacamoleo committed:
addressing review comments

guacamoleo committed:
Preparing extract_slice support for checkin

guacamoleo committed:
Adding tests and removing patch.

guacamoleo committed:
Merging 1d and 2d into combined function.

guacamoleo committed:
Adding empty rewrite function

guacamoleo committed:
refined ReduceOp working

guacamoleo committed:
preparing refine reduce for reivew

guacamoleo committed:
addressing issues and running clang-format

guacamoleo committed:
re-running clang-format on files outside this commit

ravil-mobile committed:
[AMD] fixed calculations in `fastPathComputeOffsets`

ravil-mobile committed:
[AMD] Added changes required after rebasing

ravil-mobile committed:
[AMD] Added bugfixes: canonicalizer, extract_slice

ravil-mobile committed:
[AMD] Fixed/Extended ConcatOp lowering to LLVM

ravil-mobile committed:
[AMD] Added elementwise broadcast ops refinement

ravil-mobile committed:
[AMD] Fixed MFMA to Linear Layout conversion for 1D tensors

guacamoleo commented:
Thanks Ravil!
I'm not worried about performance at this point, because we aren't really doing scheduling yet.

I see that refining buffer loads is a new function from traditional loads. We should think of a way to be able to combine many of these refinement function together to consodidate this code as much as possible. We can discuss offline. Maybe there's a way of keeping the common code in the function, but passing in a functor which does the unique operations.

guacamoleo commented:
> @guacamoleo, I tested correctness of the FA kernel with refined buffer ops; numerics is correct

Great. How about consolidating code? Is there any way to merge the loads into a single refinement function? I'm concerned about all the duplication which has been inherant in our support. If there's no way to do it right now, we'll want to address consolidating refinement code before we try to upstream this.

ravil-mobile committed:
[AMD] Adapted the code due to upstream changes

guacamoleo commented:
We discussed this offline; commit looks good after rebasing with branch.

binarman committed:
Refine pattern matching rework (#783)

This PR:

    makes Refine pass FuncOp based, instead of ModuleOp
    replaces walkers with PatternRewriters

Co-authored-by: ravil-mobile <ravil.aviva.com@gmail.com>

guacamoleo commented:
Just a reminder to look into the issue with 16-bit memory ops resulting https://github.com/ROCm/triton-internal/issues/699#issuecomment-2835306030 in conjunction with this commit.

ravil-mobile committed:
[AMD] Removed redundant headers from `RefineOps.cpp`

ravil-mobile committed:
[AMD] Inlined refine-ops functions

ravil-mobile committed:
[AMD] Added `local_alloc` refinement

guacamoleo reviewed (commented):
It looks like we're just copying axisInfo from operand[0] of extract_slice to the extract slice op itself. Is it implicit in doing this that these values get re-calculated for the smaller sizes? It seems like there are cases where we do change contiguity and divisibility. For example, if an operand was 4 vgprs which need to be allocated contiguously, then we split that in half, the new operands are only 2 vgprs which don't need to be 4-contiguous.
Am I understand or mis-understanding this?

ravil-mobile commented:
> It looks like we're just copying axisInfo from operand[0] of extract_slice to the extract slice op itself. Is it implicit in doing this that these values get re-calculated for the smaller sizes? It seems like there are cases where we do change contiguity and divisibility. For example, if an operand was 4 vgprs which need to be allocated contiguously, then we split that in half, the new operands are only 2 vgprs which don't need to be 4-contiguous. Am I understand or mis-understanding this?

Yes, you are correct. The scenario that you describe may happen. Let me think about a solution

ravil-mobile committed:
[AMD] Added a fix to elementwise refinement (#789)

- Added a lit-test to `elementwise.mlir` which test refinement of `convert_layout` from one `mma` to another `mma` layoyts
- Added a bug fix

ravil-mobile commented:
Hi @guacamoleo, I added some to recompute the `AxisInfo`. Please, verify: https://github.com/ROCm/triton/blob/240b651150c0f939b04357c03f551a925b166e2a/lib/Analysis/AxisInfo.cpp#L1033-L1090

guacamoleo commented on third_party/amd/lib/TritonAMDGPUTransforms/RefineOps.cpp:
```diff
@@ -626,6 +626,100 @@ struct LoadOpPattern : public RefineRewritePattern<triton::LoadOp> {
   }
 };
 
+struct AMDGCNBufferLoadOp
```
Can you add an explanatory comment regarding how refinement of buffer_load is more complex than that of global_load; it looks like we need to examine the refinement of masks, otherTensor and offsets and bring it all together. This'll make the function more understandable.

ravil-mobile replied:
Done

guacamoleo commented on test/Analysis/test-alignment.mlir:
```diff
@@ -866,3 +866,13 @@ tt.func @test_warp_specialize_propagation(%arg0: !tt.ptr<f16> {tt.divisibility =
+
+// -----
+
+#mma = #ttg.amd_mfma<{versionMajor = 3, versionMinor = 0, warpsPerCTA = [4, 1], instrShape = [32, 32], isTransposed = true}>
```
Thanks for adding a test!
I recall us determining that contiguity, divisitbility and constancy can all change from extract slice; can you add a test where all 3 change and we correctly test that behavior?

ravil-mobile replied:
We decided to simply propagate the AxisInfo from `extract_slice`. it is a part of the upstream

ravil-mobile committed:
[AMD] Added fixes to refine-ops-pass after rebase with upstream

ravil-mobile committed:
[AMD] lifted upstream constraints in MemDescSubview::verify

ravil-mobile committed:
[AMD] Added bufferOps refinement

guacamoleo reviewed (approved):
Thanks, Ravil. I had looked at this more closely and seen that there are multiple differences from global_loads so it doesn't make sense to try and merge this with global_load.
This looks good if tests passing and merge conflict fixed.
